R/req-perform-multi.R

```diff
@@ -0,0 +1,228 @@
+#'   If `parse_resp` is not specified, it will be a list of the raw responses.
+#' @export
+#'
+#' @examples
+#' page_size <- 150
+#'
+#' req_pokemon <- request("https://pokeapi.co/api/v2/pokemon") %>%
+#'   req_url_query(limit = page_size) %>%
+#'   req_paginate_next_url(
+#'     parse_resp = function(resp) {
+#'       parsed <- resp_body_json(resp)
+#'       results <- parsed$results
+#'       data <- data.frame(
+#'         name = sapply(results, `[[`, "name"),
+#'         url = sapply(results, `[[`, "url")
+#'       )
+#'
+#'       list(data = data, next_url = parsed$`next`)
+#'     },
+#'     n_pages = function(parsed) {
+#'       total <- parsed$count
+#'       ceiling(total / page_size)
+#'     }
+#'   )
+#'
+#' responses <- req_perform_multi(req_pokemon)
+req_perform_multi <- function(req,
+                              # paths = NULL,
+                              max_requests = NULL,
+                              cancel_on_error = FALSE,
+                              progress = TRUE,
+                              error_call = current_env()) {
+  # TODO support `path` argument
+  # * either a list of requests (same size as `n_requests`)
+  # * a path pattern
+
+  check_request(req)
+  check_has_multi_policy(req)
+  check_number_whole(max_requests, allow_infinite = TRUE, allow_null = TRUE, min = 1)
+  max_requests <- max_requests %||% Inf
+
+  n_requests <- min(req$policies$multi$n_requests, max_requests)
+  get_n_requests <- req$policies$multi$get_n_requests
+
+  # the implementation below doesn't really support an infinite amount of pages
+  # but 100e3 should be plenty
+  if (is.infinite(n_requests)) {
+    n_requests <- 100e3
+  }
+
+  perform <- error_wrapper(
+    f = req_perform,
+    cancel_on_error = cancel_on_error,
+    error_message = "When performing request {i}.",
+    error_class = "httr2_failure",
+    error_call = error_call
+  )
+
+  parse_resp <- error_wrapper(
+    f = function(resp) {
+      parse_resp <- req$policies$parse_resp
+      parse_resp(resp)
+    },
+    cancel_on_error = cancel_on_error,
+    error_message = "When parsing response {i}.",
+    error_class = "httr2_parse_failure",
+    error_call = error_call
+  )
+
+  pb <- create_progress_bar(
+    total = n_requests,
+    name = req$policies$multi$type,
+    config = progress
+  )
+  show_progress <- !is.null(pb)
+
+  out <- rep(list(cancelled_response()), n_requests)
+
+  i <- 0L
+  while ((i + 1) <= n_requests) {
+    i <- i + 1L
+
+    resp <- perform(req)
+    parsed <- parse_resp(resp)
+
+    if (i == 1) {
+      n_requests <- min(get_n_requests(parsed), n_requests)
+    }
+
+    out[[i]] <- parsed$data
+    if (show_progress) cli::cli_progress_update(total = n_requests)
+
+    req <- multi_next_request(req, parsed)
+    if (is.null(req)) {
+      break
+    }
+  }
+  if (show_progress) cli::cli_progress_done()
+
+  # remove unused tail of `out`
+  if (i < length(out)) {
+    out <- out[seq2(1, i)]
+  }
+
+  vctrs::list_unchop(out)
```

Comment:
This doesn't work well in case `cancel_on_error = FALSE` and there actually is an error. Should we only unchop if `cancel_on_error = TRUE`?

Reply:
Or only unchop the results corresponding to requests that succeeded?